src/media.rs

```diff
@@ -32,78 +29,78 @@ pub fn generate_thumbnail_and_get_size(image: &[u8]) -> (Vec<u8>, (u32, u32)) {
-
-    let file_hash = Arc::new(file_hash);
-    let callback = Arc::new(callback);
-
-    let (file_encrypted, media_key) = crypto::encrypt_media_message(media_type, file);
-    let file_encrypted_hash = crypto::sha256(&file_encrypted);
-
-
-    //Todo refactoring, remove arc -> request_file_upload fnonce
-    let file_encrypted_hash = Arc::new(file_encrypted_hash);
-    let file_encrypted = Arc::new(file_encrypted);
-    let media_key = Arc::new(media_key);
-    let file_len = file.len();
-
-    connection.request_file_upload(&file_hash.clone(), media_type, Box::new(move |url: Result<&str>| {
-        match url {
-            Ok(url) => {
-                let url = url.to_string();
-                let file_hash = file_hash.clone();
-                let file_encrypted_hash = file_encrypted_hash.clone();
-                let file_encrypted = file_encrypted.clone();
-                let media_key = media_key.clone();
-                let callback = callback.clone();
-
-                thread::spawn(move || {
-                    let form = reqwest::multipart::Form::new()
-                        .text("hash", base64::encode(&file_encrypted_hash.to_vec()))
-                        .part("file", reqwest::multipart::Part::reader(Cursor::new(file_encrypted.to_vec()))
-                            .mime(reqwest::mime::APPLICATION_OCTET_STREAM));
-
-                    let file_info = reqwest::Client::new().post(url.as_str())
-                        .multipart(form)
-                        .send()
-                        .and_then(|mut response| response.text())
-                        .map_err(|e| Error::with_chain(e, "could not upload file"))
-                        .and_then(|response| json::parse(&response).map_err(|e| (Error::with_chain(e, "invalid response"))))
-                        .and_then(|json| json.get_str("url").map(|url| url.to_string()))
-                        .map(|url| FileInfo {
-                            mime: "image/jpeg".to_string(),
-                            sha256: file_hash.to_vec(),
-                            enc_sha256: file_encrypted_hash.to_vec(),
-                            key: media_key.to_vec(),
-                            url,
-                            size: file_len, //Or encrypted file size ??
-                        });
-                    callback(file_info);
-                });
-            }
-            Err(err) => callback(Err(err).map_err(|_| "could not request file upload"))
-        }
-    }))
+pub async fn upload_file(file: &[u8], mime: String, media_type: MediaType, auth: &str, host: &str) -> Result<FileInfo> {
+     let sha256 = crypto::sha256(file);
+
+    let (file_encrypted, key) = crypto::encrypt_media_message(media_type, file);
+    let enc_sha256 = crypto::sha256(&file_encrypted);
+    let token = base64::encode_config(&enc_sha256, base64::URL_SAFE_NO_PAD);
+    let size = file.len();
+    let mut url = reqwest::Url::parse(&format!("https://{}/", host)).unwrap();
```

Comment:
Where does `host` come from, and why does it need to be surrounded in `https://` like this? What if the user already prepends `https://` before calling the function?
Maybe it'd be better to have the user pass in a `Url` object (avoiding `&str` is always good!), and provide an additional function to do the conversion into that object if necessary (alternatively, maybe the `hosts` from earlier can be parsed into URLs on reception, and an error thrown there if WA send us bogus URLs?)?
At any rate, this `unwrap` seems scary...

Reply:
The `host` comes from the `MediaConn` request. I say the `MediaConn` response should return the hosts as `Vec<url::Host>`.
https://docs.rs/url/2.1.1/url/enum.Host.html
Give me your thoughts on that.

Reply:
Sure, that sounds like what I was aiming for! Would be much appreciated :)